README.md
OCAPy
=====

OVH Client Api (in Python)

OCAPy is a python client implementing [OVH restful API](https://api.ovh.com/console/) consumption

## Disclaimer ##
**OVH is in no way involved in the development of this library**, in case of bug please use the issues tracker provided by Github

## Usage overview

```python
    # Import the main class
    from OCAPy import OCAPy
    
    # Instantiate OCAPy class
    ocapy=OCAPy(
        base_url='https://api.ovh.com/1.0/',
        app_key='YOUR APPLICATION KEY',
        app_secret='YOUR APPLICATION SECRET',
        consumer_key='YOUR CONSUMER KEY'
    )

    # And play with the API
    # A GET request: GET https://api.ovh.com/1.0/me
    request=ocapy.me.get()

    # print my city
    print request['city']

    # Want to get a specific ressource?
    # GET https://api.ovh.com/1.0/ip/213.186.33.99%2F32
    print ocapy.ip('213.186.33.99/32').get()
    
    # OK but I also want to play with POST and PUT!
    # POST https://api.ovh.com/1.0/me/ovhAccount/FR/creditOrder
    print ocapy.me.ovhAccount('FR').creditOrder.post(data={'amount':'1000'})

    # PUT https://api.ovh.com/1.0/xdsl/xdsl-xxxx-1
    print ocapy.xdsl('xdsl-xxxx-1').put(data={'description':'My XDSL description'})
    
    # And what's about DELETE?
    # DELETE https://api.ovh.com/1.0/sms/user/ocapy
    print ocapy.sms('sms-xxxx-1').user('ocapy').delete()
    
```

### Basics
1. OVH API is available [here](https://api.ovh.com/console/)
2. To get your credentials follow [this tutorial](http://www.ovh.com/fr/g934.premiers-pas-avec-l-api)
3. The code has been tested with python 2.7

### Requirements
* You only need the famous [requests > 1.0.0](http://docs.python-requests.org/en/latest/) python library
* Other libs should be available within your python installation.


### Installation
1. Download the archive
2. Unzip and go into the directory
3. run ```python setup.py install```
4. Play with it!

In other words under Linux:

```bash
wget https://github.com/pslestang/OCAPy/archive/master.zip
unzip master.zip
cd OCAPy-master/
sudo python setup.py install
```

### Configuration
Starting from version **0.2.0** OCAPy is abble to read authentication parameters from an INI configuration file. This configuration file is stored in the user's home directory.
The configuration is compounded of a main configuration part and one or several profiles part. 

**Full exemple:**

```
[ocapy]
base_url = https://api.ovh.com/1.0/
profile = full

[profile-full]
name = full
app_key = 1234AbCD5eFGh6ijk
app_secret = QQkc0c1hqVnRWbmcwY3JtdDFDeFlkYkd
consumer_key = V2ROTnkwTzllMVJaZU9odDMxQzhxcGFk
base_url = https://api.ovh.com/1.0

[profile-domains]
name = domains
app_key = 1234AbCD5eFGh6ijk
app_secret = UemxsTVZKYVpVOW9kRE14UXpoeGNRmsK
consumer_key = UVlpLWVZwVk9XOWtSRTE0VVhwb2VHTlJ
base_url = https://api.ovh.com/1.0
```

To use the authentication parameters from the configuration file, just set the option ```ocapy_profile='profile name'``` when instantiating the OCAPy class.
You may use the profile called 'default' to load the profile specified in **```[ocapy]```** section.

With the above configuration the 2 following lines are strictly the same:

```python
from OCAPy import OCAPy

# use the profile called full (profile-full)
ocapy = OCAPy(ocapy_profile='full')
# use the default profile (value set to full) 
ocapy = OCAPy(ocapy_profile='default')

```

### ocapy program
Starting from version **0.2.0** OCAPy is shiped with a program called ```ocapy``` which is for the moment a helpfull program that manage configuration file.
When using it you are able to add, delete, detail an valid a profile authentication.

The 'valid' function tests that a request on https://api.ovh.com/1.0/me is working with an authentication profile 

During 'add' process, the same request is done, if the test fails, the profile is not added.

Here the help message:

```
usage: ocapy [-h] [-c] [-s] [-p PROFILE] [-n]

optional arguments:
  -h, --help            show this help message and exit
  -c, --config          Configure OCAPy and manage profiles authentication
  -s, --shell           Start an interactive shell
  -p PROFILE, --profile PROFILE
                        Authentication profile to use
  -n, --no-color        Disable color, default is to enable unless colorama
                        library is missing
```

### API Responses
- The API responses are **decoded JSON string**
- In case of **NULL** response, **None** is returned
- In case of **HTTP error** which is also an API error, an exception is raised

```python
# A successfull request return a decoded JSON string
me=ocapy.me.get()
print "My name is: %s" % me['name']
```
    My name is: Le Stang

```python
# Ex: Adding ocapy user return a NULL response so None in python
print ocapy.sms('sms-xxxx-1').users.post(data={'login':'ocapy', 'password':'plopplop'})
```    
    None

```python    
# Ex: Adding ocapy user one more time raise an exception:"
try:
    ocapy.sms('sms-xxxx-1').users.post(data={'login':'ocapy', 'password':'plopplop'})
except Exception as e:
    print "Exception raised: %s" % e
```
    Exception raised: POST https://api.ovh.com/1.0/sms/sms-xxxx-1/users [409]: This login exists already for that account

```python
# Ex: Deleting ocapy user return a NULL response, so None in python"
print ocapy.sms('sms-xxxx-1').users('ocapy').delete()
```
    None

```python
# Calling an invalid resource, raise an exception:"
try:
    ocapy.me.invalidresource.get()
except Exception as e:
    print "Exception raised %s" % e

```
     Exception raised GET https://api.ovh.com/1.0/me/invalidresource [404]: Got an invalid (or empty) URL


### License
OCAPy is licensed under the terms of the General Public License v3


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2013-09-02)
++++++++++++++++++

* First release. 


OCAPy/input.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2013 Pierre-Samuel Le Stang (ps@lestang.fr)
#  
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>. 

COLORS = False
try:
    from colorama import init, Fore, Back, Style
    init(autoreset=True)
    COLORS = True
except ImportError:
    pass

def color(enable=False):
    global COLORS
    COLORS = enable

class UserInput(object):
    """"""
    def __init__(self, question=None, choices=None, default=None):
        self.question = question
        self.choices = choices or []
        self.default = default

    def process(self):
        if not self.question or self.question == '':
            raise OCAPyInputException("Missing question for user, review your"
                                      " code...")
        choice_string=''
        for choice in self.choices:
            if choice_string != '':
                choice_string += '/'

            if choice == self.default:
                if COLORS:
                    choice = '%s%s*%s' % (Fore.GREEN, choice,
                                         Fore.WHITE)
                else:
                    choice = '%s*' % str(choice)


            choice_string += choice

        if choice_string != '':
            choice_string = '[%s]' % choice_string

        if COLORS:
            question = '%s%s%s %s: ' % (Style.BRIGHT, self.question, Style.NORMAL, choice_string)
        else:
            question = '%s %s: ' % (self.question, choice_string)

        value = raw_input('%s' % question)

        while ( len(self.choices) and value not in self.choices ):
            if value == '':
                value = self.default

            if len(self.choices) and value not in self.choices:
                if COLORS:
                    print "%s>>>>> Answer (%s%s%s) is not valid, allowed values: %s" % (Fore.RED, Style.BRIGHT, value, Style.NORMAL, self.choices)
                else:
                    print ">>>>> Answer (%s) is not valid, allowed values: %s" % (value, self.choices)

                value = raw_input('%s' % question)

        return value

if __name__ == '__main__':
    print UserInput(question='You know what?', choices=[ 'Yes', 'No', 'I\'m happy'],
                    default='Yes').process()


OCAPy/config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2013 Pierre-Samuel Le Stang (ps@lestang.fr)
#  
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>. 

"""
.. module:: config
    :platform: Unix
    :synopsis: This module defines classes for OCAPy configuration

.. moduleauthor:: Pierre-Samuel Le Stang <ps@lestang.fr>


"""

import os
import stat
import sys
import codecs
import string
import random
import logging

from ConfigParser import SafeConfigParser

class Config(object):
    """Defines the main config class
    
    .. note::

       The configuration file is stored in the user's home directory and is called .ocapyrc

    """

    def __init__(self):
        """is called when :class:`Config` is instantiated

    .. note::

        Current user's configuration is loaded as soon as :class:`OCAPy.Config` is instantiated

        """
        self.file = '%s/.ocapyrc' % os.path.expanduser('~')
        self.profiles = []
        self.ocapy = None
        self.parser = None
        self.load()

    def add_profile(self, name=None, app_key=None, app_secret=None, section=None,
                    consumer_key=None, base_url=None):
        """create a Profile instance and add it to profiles list attribute
        
        Keyword arguments:
        :param kwargs: keywords passed to :class:`Profile` for instantiation
        :return: the added :class:`Profile` instance

        """
        profile = Profile(parser=self.parser, name=name, app_key=app_key,
                          app_secret=app_secret, consumer_key=consumer_key,
                          base_url=base_url, section=section)
        self.profiles.append(profile)
        return profile

    def delete_profile(self, name):
        profile = self.profile(name)
        if profile is not None:
            profile.delete()

    def profile(self, name):
        profile = None
        for profile in self.profiles:
            if profile.name == name:
                break
            else:
                profile = None

        return profile

    def set_default(self, profile=None):
        """set default profile name to main ocapy configuration
        
        :param profile: a Profile instance

        """
        self.ocapy.profile = profile.name

    def load(self):
        self.parser = SafeConfigParser()

        if not os.path.exists(self.file):
            with open(self.file, 'a') as f:
                f.write('')

        self.parser.read(self.file)

        for section in self.parser.sections():
            options = self.parser.options(section)
            params = {}
            for option in options:
                params[option] = self.parser.get(section, option)
            params['section'] = section

            if section == 'ocapy':
                self.ocapy = OcapyConfig(parser=self.parser, **params)
            elif str(section).startswith('profile-'):
                self.add_profile(**params)

        if self.ocapy is None:
            self.ocapy = OcapyConfig(parser=self.parser,
                                     base_url='https://api.ovh.com/1.0/')

    def save(self):
        if self.ocapy.profile == '':
            if len(self.profiles) > 0:
                self.ocapy.profile = self.profiles[0].name
        self.ocapy.save()

        for profile in self.profiles:
            profile.save()

        with open(self.file, 'w') as f:
            self.parser.write(f)

        if sys.platform.startswith('linux'):
            os.chmod(self.file, stat.S_IRUSR | stat.S_IWUSR)

class OcapyConfig(object):
    """Config parameters for Ocapy"""

    def __init__(self, profile=None, base_url=None, section=None, parser=None):
        self.section = section or 'ocapy'
        self.profile = profile or ''
        self.base_url = base_url
        self.parser = parser

        if not self.parser.has_section(self.section):
            self.parser.add_section(self.section)

    def save(self):
        self.parser.set(self.section, 'base_url', self.base_url)
        self.parser.set(self.section, 'profile', str(self.profile))

class Profile(object):
    """modelises an authentication profile"""

    def __init__(self, name=None, app_key=None, app_secret=None,
                 consumer_key=None, base_url=None, section=None, parser=None):
        self.name = name
        self.app_key = app_key
        self.app_secret = app_secret
        self.consumer_key = consumer_key
        self.base_url = base_url
        self.parser = parser
        self.section = section

        if not self.parser.has_section(self.section) and self.section is not None:
            self.parser.add_section(self.section)


    def save(self):
        if self.name is None or self.name == '':
            if self.section is not None:
                self.parser.remove_section(self.section)
            self.name = ''.join(random.choice(string.ascii_uppercase + 
                                              string.digits) for x in range(6))
            self.section = 'profile-%s' % self.name
            self.parser.add_section(self.section)
    
        if self.section is None:
            self.section = 'profile-%s' % self.name
            self.parser.add_section(self.section)

        if self.is_valid():
            if self.parser.has_section(self.section):
                self.parser.set(self.section, 'name', self.name)
                self.parser.set(self.section, 'app_key', self.app_key)
                self.parser.set(self.section, 'app_secret', self.app_secret)
                self.parser.set(self.section, 'consumer_key', self.consumer_key)
                self.parser.set(self.section, 'base_url', self.base_url)
        else:
            logging.warning('Profile not valid, removing ....')
            self.delete()

    def delete(self):
        self.parser.remove_section(self.section)

    def is_valid(self):
        from OCAPy import OCAPy
        ocapy = OCAPy(base_url=self.base_url, app_key=self.app_key,
                      app_secret=self.app_secret,
                      consumer_key=self.consumer_key)

        try:
            ocapy.me.get()
        except Exception as e:
            logging.error('Profile validation error: %s' % e)
            return False

        return True

    def __str__(self):
        name = 'Name'
        base_url = 'Base URL'
        app_key = 'Application key'
        app_secret = 'Application secret'
        consumer_key = 'Consumer key'

        maxsize = 0
        for size in [len(name), len(base_url), len(app_key),
                     len(app_secret), len(consumer_key)]:
            if size > maxsize:
                maxsize = size

        output = '%s: %s' % (string.rjust(name, maxsize), self.name)
        output += '\n%s: %s' % (string.rjust(base_url, maxsize), self.base_url)
        output += '\n%s: %s' % (string.rjust(app_key, maxsize), self.app_key)
        output += '\n%s: %s' % (string.rjust(app_secret, maxsize), self.app_secret)
        output += '\n%s: %s' % (string.rjust(consumer_key, maxsize), self.consumer_key)

        return output

if __name__ == '__main__':

    # Load config file $HOME/.ocapyrc
    config = Config()
    # Add a profile without name
    config.add_profile(base_url='https://api.ovh.com/1.0/',
                       app_key='appkey1',
                       app_secret='appsecret1',
                       consumer_key='consumerkey1',
                      )
    # Add an other one
    config.add_profile(base_url='https://api.ovh.com/1.0/',
                       app_key='appkey2',
                       app_secret='appsecret2',
                       consumer_key='consumerkey2',
                       name='name2'
                      )
    # Save the data
    # Default profile is the first profile added (unless ther is already one set)
    config.save()


OCAPy/errors.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

# Copyright 2013 Pierre-Samuel Le Stang (ps@lestang.fr)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

class OCAPyException(Exception):
    """Defines the exception base class for OCAPy classes"""

    def __init__(self, message, request=None):
        Exception.__init__(self, message)
        self.request=request

class OCAPyRequestException(OCAPyException):
    """Defines the exception for requests"""
    def __init__(self, message, request=None):
        OCAPyException.__init__(self, message)
        self.request=request

class OCAPyConfigException(OCAPyException):
    """Defines the exception class for OCAPy Config classes"""
    pass

class OCAPyInputException(OCAPyException):
    """Defines the exception class for OCAPy Config classes"""
    pass

# vim:set shiftwidth=4 tabstop=4 softtabstop=4 encoding=utf-8 expandtab textwidth=79



OCAPy/schema.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

# Copyright 2013 Pierre-Samuel Le Stang (ps@lestang.fr)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.


import requests
from errors import OCAPyException

class Schemas(object):
    """A set of Schema"""

    def __init__(self, base_url=None):
        self.base_url = base_url
        self.list = []

    def retrieve(self, format='json'):
        """retrieves all the schemas URL"""
        request = requests.get(self.base_url)

        if request.status_code != requests.codes.ok:
            raise OCAPyException('Unable to retrieve JSON schemas: %s' % 
                                 (request.json()['message']), request=request)

        for api in request.json()['apis']:
            self.list.append(Schema(base_url=self.base_url,
                                    path=str(api['schema']).format(path=api['path'],
                                    format=format),
                                    format=format))
        return self.list

    
class Schema(object):
    """modelizes a schema"""

    def __init__(self, base_url=None, path=None, format='json'):
        self.base_url = base_url
        self.path = path
        self.format = format
        self.content = None

    def load(self):
        url = '%s/%s' % (self.base_url.rstrip('/'), self.path.lstrip('/'))
        request = requests.get(url)

        if request.status_code != requests.codes.ok:
            raise OCAPyException('Unable to load JSON schema %s: %s' %
                                 (self.path, request.json()['message']), request=request)

        self.content = request.json()
        return self.content


if __name__ == '__main__':
    schemas = Schemas()
    schemas.base_url = 'https://api.ovh.com/1.0/'
    for schema in schemas.retrieve():
        print schema.path

# vim:set shiftwidth=4 tabstop=4 softtabstop=4 encoding=utf-8 expandtab textwidth=79



OCAPy/OCAPy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2013 Pierre-Samuel Le Stang (ps@lestang.fr)
#  
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>. 


import logging
import hashlib
import time
import urllib

import requests
from requests.auth import AuthBase

from config import Config
from errors import OCAPyException, OCAPyRequestException

# Current logger
logging.basicConfig(level=logging.INFO)

# Set urllib3/requests logging level to warning
logging.getLogger("requests").setLevel(logging.WARNING)
logging.getLogger("urllib3").setLevel(logging.WARNING)

# Authentication class which inherit from requests.auth.AuthBase
# requests module usage only
class OVHAuth(AuthBase):
    """OVH's API authentication schema"""

    def __init__(self, consumer_key=None, app_secret=None, app_key=None,
                 base_url=None, time_path='auth/time', url_path='/',
                 request_type='GET'):
        self.consumer_key = consumer_key
        self.app_key = app_key
        self.app_secret = app_secret
        self.base_url = base_url
        self.url_path = url_path
        self.request_type = request_type
        self.time_path = time_path
        self.time_url = '%s/%s' % (str(self.base_url).rstrip('/'),
                                   str(self.time_path).lstrip('/'))
        self.url = '%s/%s' % (str(self.base_url).rstrip('/'),
                              str(self.url_path).lstrip('/'))

    # Compute the request signature
    # Refer to http://www.ovh.com/fr/g934.premiers-pas-avec-l-api 
    def signature(self, timestamp=None):
        if timestamp is None:
            timestamp = self.now()

        sha1 = hashlib.sha1()
        sha1.update('+'.join([self.app_secret, self.consumer_key,
                              str(self.request_type).upper(), self.url, "", timestamp]))
        return '$1$'+sha1.hexdigest()

    # time is needed to sign the requests
    # Use the URL given by OVH to retrieve time
    def server_time(self):
        request = requests.get(self.time_url)
        if request.status_code != requests.codes.ok:
            raise OCAPyRequestException("Time request error: %s" % request.json()['message'], request=request)

        return int(requests.get(self.time_url).text)

    # Return the time difference between OVH and local time
    def drift_time(self):
        return self.server_time() - int(time.time())


    # Return the current time, taking care of drift time
    def now(self):
        return str(int(time.time()) + self.drift_time())

    # Make the class callable
    # rhe r argument is given by requests module
    def __call__(self, r):

        timestamp = self.now()
        r.headers['X-Ovh-Consumer'] = self.consumer_key
        r.headers['X-Ovh-Application'] = self.app_key
        r.headers['X-Ovh-Signature'] = self.signature(timestamp=timestamp)
        r.headers['X-Ovh-Timestamp'] = str(timestamp)

        return r


# Resource class described by:
# 1. name: the name of the resource (ex: server)
# 2. path: the full path to access to the resource (ex: https://api.ovh.com/1.0/dedicated/server)
# 3. api: the api object related to the resource
# 4. callable: the class to call that permit method chaining ( api.me.ovhAccount('FR').creditOrder )
class Resource(object):
    """"""

    def __init__(self, name=None, path=None, api=None, callable=None):
        self.name = name
        self.path = path
        self.api = api
        self.callable = callable

    def __call__(self, *args, **kwargs):
        # /dedicated/server/ns1234.ovh.net => dedicated.server("ns1234.ovh.net')
        if len(args) == 1:
            # quote the arg to handle such data: 'a.b.c.d/32'
            return getattr(self, urllib.quote_plus(args[0]))

        # default call is a get()
        return self.get()

    def __getattr__(self, name):
        logging.debug("getattr called for attribute %s" % name)
        path = '%s/%s' % (self.path.rstrip('/'), name)
        return self.callable(name=name,
                             api=self.api,
                             path=path,
                             callable=self.callable)

    # The common request method called by get, put, post, delete methods
    # This method return a requests instance or raise an exception if request
    # fail (return code != requests.codes.ok )
    def _request(self, type='GET', kwargs=None):

        url = '%s/%s' %(self.api.base_url.rstrip('/'), self.path.lstrip('/'))
        # method: get/post/put/delete
        method = str(type).lower()

        logging.debug("%s %s" % (method, url))
        logging.debug("path is %s" % self.path)

        # call requests
        response = getattr(requests, method)(url,
                                           auth=self.api.auth(
                                               url_path=self.path,
                                               consumer_key=self.api.consumer_key,
                                               app_key=self.api.app_key,
                                               app_secret=self.api.app_secret,
                                               base_url=self.api.base_url,
                                               request_type=method
                                           ),
                                           **kwargs)

        # check the response, and raise the exception in case of non ok HTTP code
        if response.status_code != requests.codes.ok:
            message = "%s %s [%s]: %s" % (str(type).upper(),
                                        response.url,
                                        response.status_code,
                                        response.json()['message']
                                       )
            raise OCAPyRequestException(message, request=response)
                             
        else:
            return response.json()

    def get(self, **kwargs):
        logging.debug("GET request")
        return self._request(type='GET', kwargs=kwargs)

    def put(self, **kwargs):
        logging.debug("PUT request")
        return self._request(type='PUT', kwargs=kwargs)

    def post(self, **kwargs):
        logging.debug("POST request")
        return self._request(type='POST', kwargs=kwargs)

    def delete(self, **kwargs):
        logging.debug("DELETE request")
        return self._request(type='DELETE', kwargs=kwargs)


class API(object):
    """"""

    def __init__(self, auth=None, base_url=None, app_key=None, app_secret=None,
                consumer_key=None):
        self.base_url = base_url
        self.auth = auth
        self.app_key = app_key
        self.app_secret = app_secret
        self.consumer_key = consumer_key

    # Dynamic building of resource
    def __getattr__(self, attribute):
        path='/%s' % attribute
        return Resource(name=attribute, api=self, path=path, callable=Resource)


class OCAPy(API):
    """"""
    def __init__(self, ocapy_profile=None, **kwargs):
        if ocapy_profile is not None:
            config=Config()
            if ocapy_profile == 'default':
                profile = config.profile(config.ocapy.profile)
            else:
                profile = config.profile(ocapy_profile)
                
            if profile is None:
                if ocapy_profile != 'default':
                    raise OCAPyException('OCAPy profile "%s" is unknown' % ocapy_profile)
                else:
                    raise OCAPyException('OCAPy default "%s" profile does not '
                                         'exist' % config.ocapy.profile )
            else:
                kwargs['base_url'] = profile.base_url
                kwargs['app_key'] = profile.app_key
                kwargs['app_secret'] = profile.app_secret
                kwargs['consumer_key'] = profile.consumer_key

        super(OCAPy, self).__init__(auth=OVHAuth, **kwargs)

# vim:set shiftwidth=4 tabstop=4 softtabstop=4 encoding=utf-8 expandtab textwidth=79
